### PR TITLE
Splunk OnCall migration tool

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Refactors the PagerDuty migration script to be a bit more generic + adds a migration script to migrate from Splunk OnCall (VictorOps)

tldr;
```bash
❯ docker build -t oncall-migrator .
[+] Building 0.4s (10/10) FINISHED
❯ docker run --rm \
-e MIGRATING_FROM="pagerduty" \
-e MODE="plan" \
-e ONCALL_API_URL="http://localhost:8080" \
-e ONCALL_API_TOKEN="<ONCALL_API_TOKEN>" \
-e PAGERDUTY_API_TOKEN="<PAGERDUTY_API_TOKEN>" \
oncall-migrator
running pagerduty migration script...

❯ docker run --rm \
-e MIGRATING_FROM="splunk" \
-e MODE="plan" \
-e ONCALL_API_URL="http://localhost:8080" \
-e ONCALL_API_TOKEN="<ONCALL_API_TOKEN>" \
-e SPLUNK_API_ID="<SPLUNK_API_ID>" \
-e SPLUNK_API_KEY="<SPLUNK_API_KEY>" \
oncall-migrator
migrating from splunk oncall...
```

https://www.loom.com/share/a855062d436a4ef79f030e22528d8c71

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] Added the relevant release notes label (see labels prefixed w/ `release:`). These labels dictate how your PR will
    show up in the autogenerated release notes.

### CODE REVIEW

File: tools/pagerduty-migrator/scripts/add_users_pagerduty_to_grafana.py
Comment (joeyorlando) [inline tools/pagerduty-migrator/scripts/add_users_pagerduty_to_grafana.py:1]: this has been moved to `[tools/migrators/add_users_to_grafana.py](https://github.com/grafana/oncall/pull/4267/files#diff-2b8ae690c66ef69beb40bffa232e3d9a03be89122cb0864ecae24b36ffa5156a)` (+ slightly refactored to support migrating users from Splunk OnCall -> Grafana OnCall)

File: tools/migrators/lib/tests/splunk/resources/test_schedules.py [hunk 1]
```diff
@@ -0,0 +1,903 @@
...
+        )
+
+    expected_oncall_api_create_calls_args = [
+        ("on_call_shifts", shift) for shift in expected_oncall_shift_create_calls
+    ]
+    expected_oncall_api_create_calls_args.append(
+        ("schedules", expected_oncall_schedule_create_call)
+    )
+
+    print(mock_oncall_client.create.call_args_list)
```
Comment (matiasb) [inline tools/migrators/lib/tests/splunk/resources/test_schedules.py:723]: I guess we don't need the print here (and below)?

**Review by matiasb — APPROVED** [review]
LGTM, minor questions added (and nice to see complex schedule test cases! :+1: )

File: tools/migrators/README.md [hunk 2]
```diff
@@ -1,49 +1,55 @@
...
-- User notification rules
-- On-call schedules
-- Escalation policies
-- Services (integrations)
-- Event rules (experimental, only works with global event rulesets)
+## Getting Started
 
-## Limitations
+1. Make sure you have `docker` installed and running
+2. Build the docker image: `docker build -t oncall-migrator .`
```
Comment (mderynck) [inline tools/migrators/README.md:13]: Do we want to start publishing this to save people the build step (separate issue)?

Comment (joeyorlando) [reply tools/migrators/README.md:13]: Good idea 👍 https://github.com/grafana/oncall/issues/4345

File: tools/migrators/lib/tests/splunk/resources/test_schedules.py (hunk 1 above)
Comment (joeyorlando) [reply tools/migrators/lib/tests/splunk/resources/test_schedules.py:723]: good catch, thanks! 👍

File: tools/migrators/lib/splunk/resources/paging_policies.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    notify_rule = {"user_id": user_id, "type": oncall_type, "important": False}
+
+    if not delay:
+        return [notify_rule]
+
+    wait_rule = {
+        "user_id": user_id,
+        "type": "wait",
+        "duration": transform_wait_delay(delay),
+        "important": False,
```
[the remark this answers is not in the document]
Comment (joeyorlando) [reply tools/migrators/lib/splunk/resources/paging_policies.py:58]: Correct, they have a concept called "custom paging policies", but those really don't map to how we model user notification policies, so I just decided to map Primary Paging Policy -> Default and ignore custom ones:

Screenshot 2024-05-14 at 09 24 47

(the PagerDuty migrator also hardcodes created user notification policies to `"important": False`)

File: tools/migrators/README.md [hunk 4]
```diff
@@ -188,9 +233,128 @@ If you want to include service & integration names in the names of migrated inte
...
+unmatched users or schedules that cannot be migrated won't be migrated as well.
+
+##### Caveats
+
+- delays between escalation steps may be slightly different in Grafana OnCall, see [Limitations](#limitations-1) for
+more info.
+- the following Splunk OnCall escalation step types are not supported and will not be migrated:
+  - "Notify the next user(s) in the current on-duty shift"
+  - "Notify the previous user(s) in the current on-duty shift"
+  - "Notify every member of this team"
```
[the remark this answers is not in the document]
Comment (joeyorlando) [reply tools/migrators/README.md:295]: Correct, it is. _However_, we don't currently have a script to migrate Splunk (or PagerDuty) teams to Grafana/Grafana OnCall teams (I considered adding it, and even left most of the code for it there, but commented out)

I suppose we could support this escalation step type and attempt to find the team match, and if we don't, just skip it. Maybe I'll do a follow up PR with support for this 👍

File: tools/migrators/lib/oncall/api_client.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import requests
+
+from lib.base_config import ONCALL_API_TOKEN, ONCALL_API_URL
+from lib.network import api_call as _api_call
+
+
+class OnCallAPIClient:
```
[the remark this answers is not in the document]
Comment (joeyorlando) [reply tools/migrators/lib/oncall/api_client.py:7]: I think I just wanted to keep the closely related methods together (and make the imports simpler). You're right though, there's no need for instances/initialization.

File: tools/migrators/lib/splunk/resources/schedules.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+    schedule["oncall_schedule"] = Schedule.from_dict(schedule).migrate(user_id_map)
+
+
+def _splunk_datetime_to_dt(text: str) -> datetime.datetime:
+    """
+    Convert a Splunk datetime string to a datetime object.
+    """
+    dt = datetime.datetime.strptime(text, "%Y-%m-%dT%H:%M:%SZ")
+    return dt.replace(tzinfo=datetime.timezone.utc)
```
[the remark this answers is not in the document]
Comment (joeyorlando) [reply tools/migrators/lib/splunk/resources/schedules.py:61]: I think it's okay here. They return the `timezone`, but they also return the `start` attribute in UTC. Just double checked their API response for schedule rotation shifts:
```json
{
  "label": "multi-day stuff",
  "timezone": "America/Toronto",
  "start": "2024-04-24T21:00:00Z",
  "duration": 7,
  "shifttype": "cstm",
  "periods": [],
  "current": {},
  "next": {},
  "shiftMembers": [
    {
      "username": "joeyorlando1",
      "slug": "rtm-h8lTf3VyNUAenaRXGVWD"
    },
    {
      "username": "joeyorlando",
      "slug": "rtm-QIRzjSe3jyatvpIcutLt"
    }
  ]
}
```

We use this function on the `start` attribute. This shift is set to start at 17h (America/Toronto tz).. which is 21h UTC. With that said, I suppose I don't need to do `dt.replace(tzinfo=datetime.timezone.utc)`

**Pull request merged**